app/models/user.rb

```diff
@@ -214,6 +216,70 @@ def profile_picture
     profile_picture
   end
 
+  def approve(reason, moderator)
+    ModeratorLog.create(user_id: moderator.id,
+                        target: self,
+                        action: "#{moderator.username} has approved #{username}.",
+                        reason: reason)
+    update_attribute(:approved, true)
+  end
+
+  def lift_ban(reason, moderator)
+    # Do not lift bans on users marked for death.
+    unless marked_for_death
+      site_ban = SiteBan.find_by("'#{Time.now.utc}' < expiration AND user_id = #{id}")
+      unless site_ban.nil?
+        ModeratorLog.create(user_id: moderator.id, 
+                            target: self,
+                            action: "#{moderator.username} has lifted #{username}'s ban (original expiry: #{ApplicationController.helpers.date_string_short(site_ban.expiration)}).",
+                            reason: reason)
+        site_ban.destroy
+      end
+    end
+  end
+
+  def ban_user(duration, reason, moderator)
+    # Bans cannot be modified after a user has been marked for death.
+    unless marked_for_death
+      # Find and update an existing ban.
+      site_ban = SiteBan.find_by("'#{Time.now.utc}' < expiration AND user_id = #{id}")
+      unless site_ban.nil?
+        site_ban.expiration = Time.now.utc.to_date + duration.days
+        site_ban.reason = reason
+        site_ban.save
+        ModeratorLog.create(user_id: moderator.id, 
+                            target: self,
+                            action: "#{moderator.username} has adjusted #{username}'s ban to #{duration} days.",
+                            reason: reason)
+      else
+        # Or create a new one if not found.
+        SiteBan.create(user_id: id, 
+                      expiration: Time.now.utc.to_date + duration.to_i.days, 
+                      reason: reason)
+        ModeratorLog.create(user_id: moderator.id, 
+                            target: self,
+                            action: "#{moderator.username} has banned #{username} for #{duration} days.",
+                            reason: reason)
+      end
+    end
+  end
+
+  def mark_for_death(reason, moderator)
+    ban_user(99999, reason, moderator)
+    update_attribute(:marked_for_death, true)
+    ModeratorLog.create(user_id: moderator.id, 
+                        target: self,
+                        action: "#{moderator.username} has marked #{username} for death!".upcase,
+                        reason: reason)
+    ip_addresses = UserSession.where(user_id: id).distinct.pluck(:ip_address)
+    ip_addresses.each do |ip|
+      dup_ip = IpBan.find_by(ip: ip)
+      unless dup_ip.present?
+        IpBan.create(ip: ip, reason: reason, alias: self.username)
+      end
+    end
```

Comment:
I'm a bit concerned by the use of mass IP bans in this way, because individual IPs can affect many, many people (such as, infamously, [the entire country of Qatar](https://en.wikinews.org/wiki/Qatari_proxy_IP_address_temporarily_blocked_on_Wikipedia). This sort of thing *hopefully* won't be a problem, but IP definitely != individual user, so that's something you should have in mind if things are handled this way. Definitely don't always assume that two people using the same IP means that they are alts.